### PR TITLE
Fixes #154

### LINKED ISSUE #154: Add support for Scala 3.0.0
I have a Scala 3.0.0 project with very strict compiler flags:

```
scalacOptions ++= Seq(
    "-indent",                       // Allow significant indentation.
    "-new-syntax",                   // Require `then` and `do` in control expressions.
    "-feature",                      // Emit warning and location for usages of features that should be imported explicitly.
    "-language:future",              // better-monadic-for
    "-language:implicitConversions", // Allow implicit conversions
    "-deprecation",                  // Emit warning and location for usages of deprecated APIs.
    "-Werror",                       // Fail the compilation if there are any warnings.
    "-source:future",
    (...)
)
```

The generated sources namely the `TzdbZoneRulesProvider` is causing errors such as:
```
[error] -- Error: /home/simon/Documents/SimulaEV/modules/frontend/target/scala-3.1.1/src_managed/main/java/time/zone/TzdbZoneRulesProvider.scala:16:26 
[error] 16 |  import zonedb.java.tzdb._
[error]    |                          ^
[error]    |`_` is no longer supported for a wildcard import; use `*` instead
[error]    |This construct can be rewritten automatically under -rewrite -source future-migration.
[error] -- Error: /home/simon/Documents/SimulaEV/modules/frontend/target/scala-3.1.1/src_managed/main/java/time/zone/TzdbZoneRulesProvider.scala:54:23 
[error] 54 |    val dayOfWeek = if (zor(2) >= 0) DayOfWeek.of(zor(2)) else null
[error]    |                       ^^^^^^^^^^^^^
[error]    |This construct is not allowed under -new-syntax.
[error]    |This construct can be rewritten automatically under -new-syntax -rewrite -source 3.0-migration.
```

Any suggestion on how to fix it? I see the code is generated via a sourceGenerators, maybe its possible to add another sourceGenerator that would recompile with `-new-syntax -rewrite -future-migration` (if that is even something sbt will allow).

PS: I'm trying to avoid following the [suggestion of creating a subproject just for this](https://stackoverflow.com/questions/35049913/is-it-possible-to-compile-managed-source-with-different-compiler-flags-than-unma) as it would complicate my setup a lot.

### PR DESCRIPTION
**Pull request by Lasering**
`scalacOptions` is a task (I thought it was a setting) which forced `generatedSourceDialect` to also be a task.

Tested locally in a Scala 3 (future) project and it worked, aka `generatedSourceDialect` has correctly computed. Changed the algorithm slightly: under Scala 2 with `-Xsource:3` scalac option the dialect Scala3 is chosen.

Haven't implemented any tests for this functionality. If its looking good for you I can try to implement some.

Only made the changes for ScalaJS, can easily make similar changes for JVM.

### CODE REVIEW

**Comment by cquiroz** [comment]
CI is failing because of code format, Could you check it?

**Comment by Lasering** [comment]
Forgot about scalafmt. Fixed.

File: sbt-tzdb/src/main/scala/io/github/sbt/tzdb/SbtTzdb.scala [hunk 1]
```diff
@@ -54,10 +63,20 @@ object TzdbPlugin extends AutoPlugin {
   )
   override val projectSettings    =
     Seq(
+      generatedSourceDialect := {
+        val neededFlags = Set("-indent", "-new-syntax", "-source:future")
+        if (scalaBinaryVersion.value == "3" && neededFlags.forall(scalacOptions.value.contains))
+          Dialect.Scala3Future
```
Comment (cquiroz) [inline sbt-tzdb/src/main/scala/io/github/sbt/tzdb/SbtTzdb.scala:69]: What do you mean by `Scala3Future`?

**Comment by cquiroz** [comment]
Thanks, it looks good, it would be good to get these on JVM and native.
Do you want to add some code in this PR or prefer to merge it and do another one?

File: sbt-tzdb/src/main/scala/io/github/sbt/tzdb/SbtTzdb.scala (hunk 1 above)
Comment (Lasering) [reply sbt-tzdb/src/main/scala/io/github/sbt/tzdb/SbtTzdb.scala:69]: Scala 3 code compiled with the `-source:future` flag which among other things:

- [A more specialized binding on the left hand side than the type on the right side](https://dotty.epfl.ch/docs/reference/changed-features/pattern-bindings.html#bindings-in-pattern-definitions)
- [Extending classes without the open modifier causes error](https://dotty.epfl.ch/docs/reference/other-new-features/open-classes.html#migration)
- [Old syntax for vararg splices causes error](https://dotty.epfl.ch/docs/reference/changed-features/vararg-splices.html)
- [Operator used infix without infix modifier cause errors](https://dotty.epfl.ch/docs/reference/changed-features/operators.html)
- [Wilcard arguments in Types using the old syntax (aka `_`) produce an error](https://dotty.epfl.ch/docs/reference/changed-features/wildcards.html)
- [Wildcard imports with old syntax (aka `_` and `=>`) cause errors](https://dotty.epfl.ch/docs/reference/changed-features/imports.html)

Although I conflated that with the indentation based syntax and the new control flow syntax. I can make more combinations of flags but that easily becomes unmanagable.

Comment (cquiroz) [reply sbt-tzdb/src/main/scala/io/github/sbt/tzdb/SbtTzdb.scala:69]: Right, thanks for the explantion

**Comment by cquiroz** [comment]
It is for scala-native 0.4.3. But we can just make it use regular jvm code. The one for js is just optimized

**Comment by Lasering** [comment]
The diference between the `TzdbZoneRulesProvider.scala` for jvm and native is only the use of TreeMap vs ZoneMap. So you want to use the sames files for jvm and native?

**Comment by cquiroz** [comment]
It depends, I'm not sure if `TreeMap` is available in `native`. If it is not there is no choice but to use `ZoneMap`. If there is wwe could use the same code

**Pull request merged**